### PR TITLE
Refactor publishing API presenters (part 2) 

### PR DESCRIPTION
**Pull request by mobaig**
Buliding on #2647 .

- Stop inheriting from EditionPresenter in CaseStudy, DetailedGuide,
  Publication, and DocumentCollectionPlaceholder presenters. These
  presenters were chosen because they are reasonably isolated from
  other presenters in this subsystem. The remaining presenters will be
  updated in a follow-up change.
- Extract common functionality into a set of PayloadBuilder objects that
  return specific parts of the publishing API payload for a given item.
  This allows each presenter to compose its complete payload using one or
  more of these objects, as well as attributes defined locally.
- Introduce a BaseItem presenter that provides attributes common across all
  documents sent to the publishing API.
- Extract default_update_type into a mixin.

Why?:

- Stop inheriting from EditionPresenter in CaseStudy, DetailedGuide,
  Publication, and DocumentCollectionPlaceholder presenters. These
  presenters were chosen because they are reasonably isolated from
  other presenters in this subsystem. The remaining presenters will be
  updated in a follow-up change.
- Extract common functionality into a set of PayloadBuilder objects that
  return specific parts of the publishing API payload for a given item.
  This allows each presenter to compose its complete payload using one or
  more of these objects, as well as attributes defined locally.
- Introduce a BaseItem presenter that provides attributes common across all
  documents sent to the publishing API.
- Extract default_update_type into a mixin.

Why?:

Inheritance isn't the right choice for sharing code amongst these presenters. The child classes aren't really proper subtypes of EditionPresenter or ItemPresenter. This, coupled with the 3-level class hierarchy made this part of the codebase hard to understand and difficult to change. By adopting a composition-based design instead, the contents of a given publishing API payload are much easier to determine.

### CODE REVIEW

File: app/presenters/publishing_api_presenters/payload_builder/access_limitation.rb [hunk 1]
```diff
@@ -0,0 +1,11 @@
+module PublishingApiPresenters
+  module PayloadBuilder
+    class AccessLimitation
+      def self.for(item)
+        return {} unless item.access_limited? && !item.publicly_visible?
+        users = User.where(organisation: item.organisations)
+        { access_limited: { users: users.map(&:uid).compact } }
```
Comment (boffbowsh) [inline app/presenters/publishing_api_presenters/payload_builder/access_limitation.rb]: `users.pluck(:uid).compact`

Comment (mobaig) [inline app/presenters/publishing_api_presenters/payload_builder/access_limitation.rb]: Updated!

File: app/presenters/publishing_api_presenters/detailed_guide.rb [hunk 2]
```diff
@@ -1,37 +1,76 @@
...
-  include PublishingApiPresenters::PoliticalHelper # Detailed Guides need a government to publish successfully.
+class PublishingApiPresenters::DetailedGuide
+  include PublishingApiPresenters::UpdateTypeHelper
+
+  attr_accessor :item
+  attr_accessor :update_type
+
+  def initialize(item, update_type: nil)
+    self.item = item
+    self.update_type = update_type || default_update_type(item)
```
Comment (jackscotti) [inline app/presenters/publishing_api_presenters/detailed_guide.rb]: why are you using `self` instead of `@`?

Comment (mobaig) [inline app/presenters/publishing_api_presenters/detailed_guide.rb]: pre-existing code, propagating the same style for consistency.

**Comment by boffbowsh** [comment]
👍

**Pull request merged**